jerry-core/ecma/operations/ecma-atomics-object.c

```diff
@@ -0,0 +1,247 @@
+ * limitations under the License.
+ */
+
+#include "ecma-atomics-object.h"
+#include "ecma-arraybuffer-object.h"
+#include "ecma-shared-arraybuffer-object.h"
+#include "ecma-bigint.h"
+#include "ecma-typedarray-object.h"
+#include "ecma-objects.h"
+#include "ecma-builtins.h"
+#include "ecma-exceptions.h"
+#include "ecma-gc.h"
+#include "ecma-globals.h"
+#include "ecma-helpers.h"
+#include "jmem.h"
+#include "jcontext.h"
+#include "ecma-function-object.h"
+
+#if JERRY_BUILTIN_ATOMICS
+
+/** \addtogroup ecma ECMA
+ * @{
+ *
+ * \addtogroup ecmaatomicsobject ECMA builtin Atomics helper functions
+ * @{
+ */
+
+/**
+ * Atomics validate Shared integer typedArray
+ *
+ * See also: ES11 24.4.1.1
+ *
+ * @return ecma value
+ */
+ecma_value_t
+ecma_validate_shared_integer_typedarray (ecma_value_t typedarray,
+                                         bool waitable)
+{
+  /* 2. */
+  if (!ecma_is_value_object (typedarray))
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Argument 'this' is not an object"));
+  }
+
+  /* 3-4. */
+  ecma_object_t *typedarray_p = ecma_get_object_from_value (typedarray);
+  ecma_typedarray_info_t target_info =  ecma_typedarray_get_info (typedarray_p);
+
+  /* 5-6. */
+  if (waitable)
+  {
+    if (!(target_info.id == ECMA_BIGINT64_ARRAY || target_info.id == ECMA_INT32_ARRAY))
+    {
+      return ecma_raise_type_error (ECMA_ERR_MSG ("Argument is not supported"));
+    }
+  }
+  else
+  {
+    if (target_info.id  == ECMA_UINT8_CLAMPED_ARRAY || target_info.id == ECMA_FLOAT32_ARRAY
+        || target_info.id == ECMA_FLOAT64_ARRAY)
+    {
+      return ecma_raise_type_error (ECMA_ERR_MSG ("Argument is not supported"));
+    }
+  }
+
+  /* 7. */
+  JERRY_ASSERT (target_info.array_buffer_p != NULL);
+
+  /* 8-10. */
+  ecma_object_t *buffer = ecma_typedarray_get_arraybuffer (typedarray_p);
+
+  if (!ecma_object_class_is (buffer,ECMA_OBJECT_CLASS_SHARED_ARRAY_BUFFER))
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Argument is not SharedArrayBuffer"));
+  }
+
+  return ecma_make_object_value (buffer);
+} /* ecma_validate_shared_integer_typedarray */
+
+/**
+ * Atomics validate Atomic Access
+ *
+ * See also: ES11 24.4.1.2
+ *
+ * @return ecma value
+ */
+ecma_value_t
+ecma_validate_atomic_access (ecma_value_t typedarray,
+                             ecma_value_t requestIndex)
+{
+  /* 1. */
+  JERRY_ASSERT (ecma_is_value_object (typedarray)
+  && ecma_typedarray_get_arraybuffer (ecma_get_object_from_value (typedarray)) != NULL);
+
+  ecma_object_t *typedarray_p = ecma_get_object_from_value (typedarray);
+
+  /* 2. */
+  ecma_number_t accessIndex;
+  ecma_op_to_index (requestIndex, &accessIndex);
+
+  /* 3. */
+  ecma_typedarray_info_t target_info =  ecma_typedarray_get_info (typedarray_p);
+  ecma_value_t length = target_info.length;
+
+  /* 4. */
+  JERRY_ASSERT (accessIndex >= 0);
+
+  /* 5-6. */
+  return (accessIndex >= length ? ecma_raise_range_error (ECMA_ERR_MSG ("Invalid length"))
```

Comment:
`ecma_number_to_uint32` returns with raw u32, however raise_range_error returns with an ecma_value. These are not compatible.